### PR TITLE
e2e test init

### PR DESCRIPTION
**Pull request by Sczlog**
1. update builtin for e2e test
2. add e2e isolated world class
3. make some example for keyboard and mouse event.

### CODE REVIEW

**Comment by Sczlog** [comment]
@netweng PTAL

File: __tests__/web/isolatedWorld.ts [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        if (newRes.result.subtype === 'error') {
+          throw new Error(newRes.result.value);
+        }
+        curr.res = newRes;
+        return true;
+      },
+    },
+  );
+
+export class IsolatedWorld {
```
Comment (netweng) [inline __tests__/web/isolatedWorld.ts:34]: add detail comments for this class purpose

File: __tests__/web/isolatedWorld.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+      resolver = res;
+    });
+
+    this.recorderInjected = new Promise<boolean>((res) => {
+      this.recorderResolver = res;
+    });
+
+    const handler = (event: Protocol.Runtime.ExecutionContextCreatedEvent) => {
+      const { id, name } = event.context;
+      if (name === '__puppeteer_utility_world__') {
```
Comment (netweng) [inline __tests__/web/isolatedWorld.ts]: __puppeteer_utility_world__  should be constant

File: __tests__/web/isolatedWorld.ts (hunk 1 above)
Comment (netweng) [reply __tests__/web/isolatedWorld.ts:34]: should explain why we need to get the isloation world

File: jest-puppeteer.config.js [hunk 3]
```diff
@@ -2,5 +2,6 @@ module.exports = {
   launch: {
     headless: true,
     product: 'chrome',
-  }
+    devtools: false
```
Comment (netweng) [inline jest-puppeteer.config.js]: launch without devtools will be faster?

File: __tests__/web/keyboard/index.spec.ts [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+    await isolatedWorld.useFakeDateTime(100);
+    await isolatedWorld.createRecorder();
+  });
+
+  it('should match texting event', async () => {
+    await (await page.$('input'))?.focus();
+    await page.keyboard.type('12345');
+    // expect record keyboard event and match type texting;
+    const curr = await isolatedWorld.retrieveCurrentEvent();
+    expect(curr.events).toMatchInlineSnapshot(`
```
Comment (netweng) [inline __tests__/web/keyboard/index.spec.ts]: individual snapshot file would be better for review

File: __tests__/web/keyboard/index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+      Array [
+        Object {
+          "code": "Digit1",
+          "key": "1",
+          "keyCode": 49,
+          "modifiers": Object {},
```
Comment (netweng) [inline __tests__/web/keyboard/index.spec.ts]: modifier value is empty  matches expection ?

File: __tests__/web/keyboard/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+          "timestamp": 2500,
+          "type": "keyup",
+        },
+      ]
+    `);
+    expect(curr.type).toMatchInlineSnapshot(`"TEXT"`);
+    expect(curr.target).toMatchInlineSnapshot(
+      {
+        id: expect.any(String),
```
Comment (netweng) [inline __tests__/web/keyboard/index.spec.ts]: what is this for?

File: __tests__/web/keyboard/index.spec.ts [hunk 7]
```diff
@@ -0,0 +1,495 @@
+import { Page } from 'puppeteer/lib/types';
+import { IsolatedWorld } from '../isolatedWorld';
+
+describe('keyboard', () => {
```
Comment (netweng) [inline __tests__/web/keyboard/index.spec.ts:6]: cool. and we also need copy-and-paste case ?

File: package.json [hunk 8]
```diff
@@ -13,7 +13,8 @@
     "prepublish": "npm run build && npm run typings",
     "lint": "prettier --write src/* && eslint src --ext .ts",
     "test:unit": "jest --testPathPattern=__tests__/units --passWithNoTests",
-    "test:web": "jest --testPathPattern=__tests__/web --preset=jest-puppeteer"
+    "test:web": "jest --testPathPattern=__tests__/web --preset=jest-puppeteer",
```
Comment (netweng) [inline package.json:16]: can we jest use one script with `--runInBand` ?

File: src/builtin.ts [hunk 9]
```diff
@@ -16,10 +16,47 @@ export type InteractionRecorderOptions = {
   onUpdateStep: (step: MatcherStep) => void;
 };
 
-class ElementSerializer {
+export class ElementSerializer {
```
Comment (netweng) [inline src/builtin.ts:19]: need detail comments

File: jest-puppeteer.config.js (hunk 3 above)
Comment (Sczlog) [reply jest-puppeteer.config.js]: devtool cannot set to true when headless is true

File: __tests__/web/keyboard/index.spec.ts (hunk 5 above)
Comment (Sczlog) [reply __tests__/web/keyboard/index.spec.ts]: becaues no modified key is pressed when typing, i will add some test with modifier soon

File: __tests__/web/keyboard/index.spec.ts (hunk 6 above)
Comment (Sczlog) [reply __tests__/web/keyboard/index.spec.ts]: id is random() and cannot predict, tell jest not to match this property when comparing snapshot.

File: __tests__/web/keyboard/index.spec.ts (hunk 7 above)
Comment (Sczlog) [reply __tests__/web/keyboard/index.spec.ts:6]: yes.

File: package.json (hunk 8 above)
Comment (Sczlog) [reply package.json:16]: i miss understood it. I will add an alter script for it

File: __tests__/web/keyboard/index.spec.ts (hunk 5 above)
Comment (netweng) [reply __tests__/web/keyboard/index.spec.ts]: should it has default value :  `{ metaKey: false, shiftKey: false ，ctrlKey: false, altKey: false }` ？

Comment (Sczlog) [reply __tests__/web/keyboard/index.spec.ts]: I will update toModifier function to remove or operator
In another pr?

**Comment by Sczlog** [comment]
@netweng PTAL

**Pull request merged**